lib/ansible/module_utils/fortios.py

```diff
@@ -198,3 +208,819 @@ def get_error_infos(self, cli_errors):
+            update_config = self._update_config
+        current_config, update_config, existing_objects_only_in_update = self._setup_configs_for_diff(
+            current_config, update_config)
+
+        if existing_objects_only_in_update:
+            return False
+
+        if isinstance(update_config, dict):
+            if not self._dictionaries_match(current_config, update_config):
+                return False
+
+        elif isinstance(update_config, list):
+            if not self._lists_match(current_config, update_config):
+                return False
+
+        return True
+
+    def _setup_configs_for_diff(self, current_config, update_config):
+        if isinstance(current_config, dict):
+            objects_only_in_update_config = {}
+        else:
+            update_object_ids = [o[self._object_identifier]
+                                 for o in update_config]
+            check_current_matches = [o for o in current_config if o[
+                self._object_identifier] in update_object_ids]
+            current_match_ids = [o[self._object_identifier]
+                                 for o in check_current_matches]
+            objects_only_in_update_config = [o for o in update_config if o[self._object_identifier] not in current_match_ids and
+                                             o[self._object_identifier] not in self._permanent_object_ids]
+            if update_config:
+                current_config = [self._get_object_from_list_by_id(
+                    current_config, obj_id) for obj_id in update_object_ids]
+
+        return current_config, update_config, objects_only_in_update_config
+
+    def _unknown_types_match(self, current_config, update_config):
+        # the _match functions determine whether every single key and value in
+        # a given pair of objects match recursively.
+        if current_config is None:
+            return False
+
+        if isinstance(update_config, dict):
+            if not self._dictionaries_match(current_config, update_config):
+                return False
+        elif isinstance(update_config, list):
+            if not self._lists_match(current_config, update_config):
+                return False
+        else:
+            if current_config != update_config:
+                return False
+
+        return True
+
+    def _dictionaries_match(self, current_config, update_config):
+        for key, val in update_config.items():
+            current_val = current_config.get(key)
+
+            if isinstance(val, dict):
+                if not self._dictionaries_match(current_val, val):
+                    return False
+
+            elif isinstance(val, list):
+                if not self._lists_match(current_val, val):
+                    return False
+
+            elif val != current_val and key not in self._match_ignore_params:
+                return False
+
+        return True
+
+    def _lists_match(self, current_config, update_config):
```

Comment:
I'm slightly confused by why functions like `_lists_match`, `_dictionaries_match` and `_unknown_types_match` are needed.

Python can already natively compare these things. Is it due to order?